opencog/rule-engine/forwardchainer/DefaultForwardChainerCB.cc

```diff
@@ -46,76 +46,140 @@ DefaultForwardChainerCB::DefaultForwardChainerCB(AtomSpace& as,
+            return derived_rules;
+
+        AtomSpace temp_pm_as;
+        Handle hcpy = temp_pm_as.add_atom(target);
+        Handle implicant_vardecl = temp_pm_as.add_atom(
+                gen_sub_varlist(target, rule->get_vardecl()));
+        Handle sourcecpy = temp_pm_as.add_atom(source);
+
+        BindLinkPtr bl =
+        createBindLink(HandleSeq { implicant_vardecl, hcpy, hcpy });
+        VarGroundingPMCB gcb(&temp_pm_as);
+        gcb.implicand = bl->get_implicand();
+
+        bl->imply(gcb);
+
+        auto del_by_value =
+                [] (std::vector<std::map<Handle,Handle>>& vec_map,const Handle& h) {
+                    for (auto& map: vec_map)
+                       for(auto& it:map) { if (it.second == h) map.erase(it.first);}
+                };
+
+        //We don't want implicant_var list to be matched as
+        //in the case of free vars in modus-ponens rules.
+        del_by_value(gcb.term_groundings, implicant_vardecl);
+        del_by_value(gcb.var_groundings, implicant_vardecl);
+
+        FindAtoms fv(VARIABLE_NODE);
+        for (const auto& termg_map : gcb.term_groundings) {
+            for (const auto& it : termg_map) {
+                if (it.second == sourcecpy) {
+                    fv.search_set(it.first);
+
+                    Handle rhandle = rule->get_handle();
+                    HandleSeq new_candidate_rules = substitute_rule_part(
+                            temp_pm_as, temp_pm_as.add_atom(rhandle), fv.varset,
+                            gcb.var_groundings);
+
+                    for (Handle nr : new_candidate_rules) {
+                        if (find(derived_rules.begin(), derived_rules.end(), nr) == derived_rules.end()) {
+                            //Adding back to _as avoids UUID clashes.
+                            Handle h = _as.add_atom(nr);
+                            //Avoid adding original rule to derived rule list
+                            if (h != rhandle)
+                                derived_rules.push_back(h);
+                        }
                     }
                 }
             }
         }
+    }
+
+    return derived_rules;
+}
+/**
+ *  Tries to unify sub atoms of implicant lists in @param rule with @param source.
+ *
+ *  @param source  An atom that might bind to variables in @param rule.
+ *  @param rule    The rule object whose implicants are to be sub atom unified.
+ *
+ *  @return        HandleSeq of derived rules from sub atoms unification.
+ */
+HandleSeq DefaultForwardChainerCB::subatom_unify(Handle source,Rule* rule)
+{
+    HandleSeq derived_rules;
+    HandleSeq impl_members = rule->get_implicant_seq();
+
+    UnorderedHandleSet output_expanded;
+    for (Handle h : impl_members) {
+        UnorderedHandleSet hs = get_all_unique_atoms(h);
+        hs.erase(h); //Already tried to unify this.
+
+        HandleSeq result = unify(source, rule);
+        derived_rules.insert(derived_rules.begin(), result.begin(),
```

Comment:
`output_expanded` is unused.  It looks like you will unify the same thing multiple times when the same atom appear inside multiple clause.